src/main/java/com/iabtcf/v2/BitVectorTCModelV2.java

```diff
@@ -19,296 +21,364 @@
-        this.consentLanguage = readStr2(bv, CONSENT_LANGUAGE_OFFSET);
-        this.vendorListVersion = bv.readBits12(VENDOR_LIST_VERSION_OFFSET);
-        this.policyVersion = bv.readBits6(TCF_POLICY_VERSION_OFFSET);
-        this.isServiceSpecific = bv.readBits1(IS_SERVICE_SPECIFIC_OFFSET);
-        this.useNonStandardStacks = bv.readBits1(USE_NON_STANDARD_STACKS_OFFSET);
-        this.specialFeaturesOptInts =
-                fillSet(SPECIAL_FEATURE_OPT_INS_OFFSET, SPECIAL_FEATURE_OPT_INS_LENGTH, bv);
-        this.purposesConsent = fillSet(PURPOSES_CONSENT_OFFSET, PURPOSES_CONSENT_LENGTH, bv);
-        this.purposesLITransparency =
-                fillSet(PURPOSE_LI_TRANSPARENCY_OFFSET, PURPOSE_LI_TRANSPARENCY_LENGTH, bv);
-        this.isPurposeOneTreatment = bv.readBits1(PURPOSE_ONE_TREATMENT_OFFSET);
-        this.publisherCountryCode = readStr2(bv, PUBLISHER_CC_OFFSET);
-
-        int currentPointer = PUBLISHER_CC_OFFSET + (CHAR.length() * 2); // publisher cc offset
-        this.vendorConsents = new TreeSet<>();
-        currentPointer = this.fetchSet(this.vendorConsents, currentPointer, bv);
-
-        this.vendorLegitimateInterests = new TreeSet<>();
-        currentPointer = this.fetchSet(this.vendorLegitimateInterests, currentPointer, bv);
-
-        this.publisherRestrictions = new ArrayList<>();
-        this.fillPublisherRestrictions(publisherRestrictions, currentPointer, bv);
-
-        this.disclosedVendors = new TreeSet<>();
-        this.allowedVendors = new TreeSet<>();
-        this.publisherPurposesConsent = new TreeSet<>();
-        this.publisherPurposesLITransparency = new TreeSet<>();
-        this.customPurposesConsent = new TreeSet<>();
-        this.customPurposesLITransparency = new TreeSet<>();
-    }
-
-    private BitVectorTCModelV2(ByteBitVector coreVector, ByteBitVector... theRest) {
-        this(coreVector);
-        for (ByteBitVector bitVector : theRest) {
-            fillRemainingVector(bitVector);
-        }
-    }
-
-    public static BitVectorTCModelV2 fromBitVector(
-            ByteBitVector coreBitVector, ByteBitVector... remainingVectors) {
+    private int version;
+    private Instant consentRecordCreated;
+    private Instant consentRecordLastUpdated;
+    private int consentManagerProviderId;
+    private int consentManagerProviderVersion;
+    private int consentScreen;
+    private String consentLanguage;
+    private int vendorListVersion;
+    private int policyVersion;
+    private boolean isServiceSpecific;
+    private boolean useNonStandardStacks;
+    private Set<Integer> specialFeaturesOptInts;
+    private Set<Integer> purposesConsent;
+    private Set<Integer> purposesLITransparency;
+    private boolean isPurposeOneTreatment;
+    private String publisherCountryCode;
+    private Set<Integer> vendorConsents;
+    private Set<Integer> vendorLegitimateInterests;
+    private List<PublisherRestriction> publisherRestrictions;
+    private Set<Integer> disclosedVendors;
```

Comment:
Thoughts on defaulting these and other optional segment fields to empty collections so that we don't return null when the segments they are parsed from aren't present?

Reply:
Mis spoke. We don’t return null but do create a new set even if the segment isn’t there. Could default to Collections.emptySet() and only create a new set when we have the optional segment. Along with the HashSet comment, not needed now. Just another small optimization that can be done.

Reply:
for consistency, i've left the field initialization to the methods.